src/apispec/core.py

```diff
@@ -18,6 +20,54 @@
 VALID_METHODS = {2: VALID_METHODS_OPENAPI_V2, 3: VALID_METHODS_OPENAPI_V3}
 
 
+def get_ref(obj_type, obj, openapi_major_version):
+    """Return object or rererence
+
+    If obj is a dict, it is assumed to be a complete description and it is returned as is.
+    Otherwise, it is assumed to be a reference name as string and the corresponding $ref
+    string is returned.
+
+    :param str obj_type: "parameter" or "response"
+    :param dict|str obj: parameter or response in dict form or as ref_id string
+    :param int openapi_major_version: The major version of the OpenAPI standard
+    """
+    if isinstance(obj, dict):
+        return obj
+    return build_reference(obj_type, openapi_major_version, obj)
+
+
+def clean_parameters(parameters, openapi_major_version):
+    """Ensure that all parameters with "in" equal to "path" are also required
+    as required by the OpenAPI specification, as well as normalizing any
+    references to global parameters and checking for duplicates parameters
+
+    See https ://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.2.md#parameterObject.
+
+    :param list parameters: List of parameters mapping
+    :param int openapi_major_version: The major version of the OpenAPI standard
+    """
+    for parameter in parameters:
+        if isinstance(parameter, dict) and parameter.get("in") == "path":
+            parameter["required"] = True
+
+    # OpenAPI Spec 3 and 2 don't allow for duplicated parameters
+    # A unique parameter is defined by a combination of a name and location
+    try:
+        param_objs = [(p["name"], p["in"]) for p in parameters if isinstance(p, dict)]
+    except KeyError as e:
+        raise InvalidParameterError("Missing key {} for parameter".format(e))
+
+    seen = set()
+    for p in param_objs:
+        if p in seen:
+            raise DuplicateParameterError(
+                "Duplicate parameter with name {} and location {}".format(p[0], p[1])
+            )
+        seen.add(p)
+
```

Comment:
I'm pretty sure we could rework this function to merge the loops.

Let's give it a try.

```py
seen = set()
for parameter in [p for p in parameters if isinstance(p, dict)]:

    # Check missing name/location
    missing_attrs = [attr for attr in ("name", "in") if attr not in parameter]
    if missing_attrs:
        raise InvalidParameterError("Missing keys {} for parameter".format(missing_attrs))

    # OpenAPI spec doesn't allow for duplicated parameters
    # A unique parameter is defined by a combination of a name and location
    unique_key = (parameter["name"], parameter["in"])
    if unique_key in seen:
        raise DuplicateParameterError(
            "Duplicate parameter with name {} and location {}".format(parameter["name"], parameter["in"])
        )
    seen.add(unique_key)

    # Add "required" attribute to path parameters
    if parameter["in"] == path:
        parameter["required"] = True
```

Do you like this implementation better?

BTW, we should probably also force `required = True` for path parameters in `Components.parameter` so that it is enforced on parameter references as well. Don't feel like you have to do it in this PR, but if you want to tackle this, no problem.

Reply:
Good catch, I've didn't look to the whole picture after latest update but I agree with this implementation, will make the update

Reply:
done in a2069bd296cef28aa4f51fe169a11ce06be26c44

Reply:
Enforced `required` in `Components.parameter` in a2069bd296cef28aa4f51fe169a11ce06be26c44